In [0]:
file_location = "/FileStore/tables/NYSE_sentiment_v4_5__continuous_y_.csv"
file_type = "csv"
 
# CSV options
infer_schema = "true"
first_row_is_header = "true"
delimiter = ","
 
df = spark.read.format(file_type) \
  .option("inferSchema", infer_schema) \
  .option("header", first_row_is_header) \
  .option("sep", delimiter) \
  .load(file_location)
 
 
display(df)

Volume ROC_20 EMA_200 Oil Gold Copper-F Wheat-F London_exg German_exg-F Japan_exg-F Dollar_index Canadian_dlr British_Pound Japan_Yen Chinese_Yuan NYSE NYSE_sentiment 0.152 4.4954 1120.4 -0.0179 -0.0042 0.66 -0.3 -0.0042 0.35 -0.52 0.51 2.0E-5 7.1E-6 4.0E-4 -1.62E-5 -0.003443906 neutral -0.2227 3.6755 1121.1 0.0225 -2.0E-4 0.42 -2.05 0.0069 0.55 -0.32 -0.14 2.0E-5 7.1E-6 4.0E-4 -1.62E-5 0.00671489 bullish 0.2584 2.2917 1121.5 -0.042 -0.0207 -2.47 -2.58 -0.0067 -0.44 0.42 1.62 0.0156 -0.012 0.0042 6.03E-5 -0.019475399 bearish -0.1022 3.8694 1122.1 0.0297 0.0 0.88 2.52 0.0044 0.61 -1.57 -1.29 -0.01036 0.00905 -0.006 8.89E-4 0.013492037 bullish -0.08 4.9279 1122.6 -0.0232 0.0034 -0.32 -3.09 0.0051 1.16 -0.11 0.32 0.00352 -0.00841 0.003 -2.56E-4 -0.001081917 neutral -0.313 2.9956 1123.2 0.014 -0.0156 0.42 -0.04 -0.0029 0.02 0.53 0.06 2.0E-5 7.1E-6 4.0E-4 -1.62E-5 9.63352E-4 neutral 0.3284 3.8051 1123.9 0.0105 0.0113 1.71 0.09 0.0018 0.54 0.11 -0.48 2.0E-5 7.1E-6 4.0E-4 -1.62E-5 0.003119767 bullish -0.0042 3.3057 1124.5 0.0073 -0.006 0.18 3.65 -0.0078 -0.25 -0.74 0.79 0.00412 0.00894 0.0063 4.96E-4 -0.002064576 neutral 0.0314 3.2951 1125.1 -0.0085 -0.0038 -2.36 0.61 -0.0107 -0.8 0.53 0.57 0.00434 -0.00479 0.0042 0.00269 -0.006630064 neutral -0.012 3.7312 1125.6 0.0037 0.0068 0.32 2.06 0.0057 0.48 -0.43 -1.07 0.0 0.0 0.0 0.00106 0.003205507 bullish -0.1741 3.2297 1126.2 -0.0091 0.0099 -1.4 -0.4 -5.0E-4 0.12 -1.92 -0.05 2.0E-5 7.1E-6 4.0E-4 -1.62E-5 0.001132601 neutral 0.1671 4.1644 1126.8 0.0183 0.0058 1.26 -2.82 0.0034 -0.11 -0.33 0.04 2.0E-5 7.1E-6 4.0E-4 -1.62E-5 -5.51037E-4 neutral -0.0637 2.8275 1127.5 0.0117 -0.0026 1.43 -1.56 0.011 0.78 0.11 -0.75 -0.0053 7.38E-4 0.0019 0.00333 0.009712097 bullish 0.2067 3.2751 1128.2 0.0064 -0.0041 -1.38 -0.11 -0.0015 -0.7 1.39 -0.31 -0.00385 -0.00172 0.0014 -0.00214 0.003464724 bullish 0.2208 5.4401 1129.1 0.0242 0.0264 3.33 3.67 0.0198 1.82 0.67 -0.78 -4.0E-4 0.00465 0.0047 -5.1E-4 0.022872061 very bullish -0.0102 5.2096 1130.0 0.0042 0.0105 0.97 1.18 0.0021 0.27 3.1 0.88 -0.00592 0.0105 -0.0034 -0.00194 0.002342453 bullish -0.3016 4.9712 1131.0 0.0025 -0.005 0.23 1.1 -0.0043 0.07 0.73 0.63 -0.00262 -0.00408 0.0045 -9.77E-4 -0.002366461 neutral 0.2313 3.7298 1131.8 -3.0E-4 0.0234 2.2 -2.27 0.0043 0.53 -0.31 0.53 0.0045 -0.00575 -6.0E-4 0.00382 -0.010265752 neutral -0.0591 3.4471 1132.7 0.0084 -0.0215 -1.83 -1.53 -0.0099 -0.96 1.24 0.25 0.00259 -0.00722 0.0061 -0.00565 0.005861865 bullish -0.1382 3.3847 1133.5 0.0 0.0058 1.31 -0.7 -3.0E-4 0.02 0.51 0.76 -0.00665 0.00764 0.0076 -0.00136 -0.003126151 neutral 0.0719 1.9572 1134.1 -0.0328 -0.0072 -3.21 -5.04 -0.0031 0.1 -1.22 -0.18 0.004 -0.00103 0.0025 -0.00143 -0.012947933 neutral -0.1686 1.1007 1134.8 -1.0E-4 -0.0144 0.87 0.19 0.0041 0.75 0.82 0.56 0.00617 0.00145 7.0E-4 0.002 -7.5171E-4 neutral 0.4604 1.067 1135.2 -0.03 -0.0142 -5.07 -6.53 -0.0238 -1.66 -0.41 0.88 -6.9E-4 -0.00533 0.0084 7.39E-4 -0.019019324 neutral -0.2368 0.0356 1135.6 -0.0231 -0.0085 0.11 1.0 0.0019 0.43 0.51 -0.16 0.01198 -0.00972 0.0026 -5.43E-4 0.002158528 bullish 0.2004 1.3921 1136.2 0.018 0.0095 2.74 2.09 0.0134 1.84 1.53 -0.58 0.00127 0.00213 -0.0034 5.73E-4 0.017516943 bullish -0.2159 1.4073 1136.9 -0.0028 -0.0057 0.08 5.63 -0.0062 0.21 0.7 -0.15 -0.00469 0.00822 0.0053 -0.00149 0.002774318 bullish 0.0038 1.0307 1137.5 -0.005 0.0104 -2.19 0.31 -0.0091 -0.22 0.7 0.23 -0.00226 -0.00296 -7.0E-4 6.95E-4 -0.004028262 neutral 0.1202 -0.4141 1137.9 -0.0055 0.0155 -1.28 -0.84 -0.0175 -1.59 -1.85 1.27 0.00321 -0.0036 -0.0023 6.49E-4 -0.018334335 neutral -0.1812 1.3447 1138.5 0.03 -0.0036 1.49 1.01 0.0136 1.55 1.08 0.24 0.0022 -0.00888 -3.0E-4 3.02E-4 0.014521896 bullish -0.5231 0.4747 1139.0 0.0079 -0.0128 -0.32 0.16 0.0021 -0.41 -0.2 0.61 -0.01253 -0.00287 0.0043 7.84E-4 -0.010386203 neutral 1.2762 0.3803 1139.5 0.0222 0.0015 0.19 0.69 -0.0208 -1.99 0.6 0.58 0.01001 -0.0108 0.0063 0.00279 -0.002293194 neutral 0.1664 -0.3234 1139.9 -0.0188 0.0195 1.73 -0.25 -0.0041 -0.28 -

In [0]:
import pyspark.sql.functions as f

In [0]:
 df.dtypes 

Out[53]: [('Volume', 'double'),
 ('ROC_20', 'double'),
 ('EMA_200', 'double'),
 ('Oil', 'double'),
 ('Gold', 'double'),
 ('Copper-F', 'double'),
 ('Wheat-F', 'double'),
 ('London_exg', 'double'),
 ('German_exg-F', 'double'),
 ('Japan_exg-F', 'double'),
 ('Dollar_index', 'double'),
 ('Canadian_dlr', 'double'),
 ('British_Pound', 'double'),
 ('Japan_Yen', 'double'),
 ('Chinese_Yuan', 'double'),
 ('NYSE', 'double'),
 ('NYSE_sentiment', 'string')]

In [0]:

#permanent_table_name = "NYSE_sentiment_perm3"
 
#df.write.format("parquet").saveAsTable(permanent_table_name)

Queries: Techinal Indicator insights

In [0]:
%sql 
 
-- highest Trading Volume percentage growth per NYSE sentiment class
 
select max(Volume) as Highest_Vol_Perc_grwth, NYSE_sentiment 
from NYSE_sentiment_perm3
group by NYSE_sentiment
order by Highest_Vol_Perc_grwth 
desc limit 5

In [0]:
%sql 
 
--average Trading Volume growth percentage when NYSE is Very Bullish 
 
select avg(Volume)*(100) as Avg_Volume_Perc_Growth 
from NYSE_sentiment_perm3
where NYSE_sentiment = "very bullish"

In [0]:
%sql 
 
--average Trading Volume growth percentage when NYSE is Very Bullish 
 
select avg(Volume)*(100) as Avg_Volume_Perc_Growth 
from NYSE_sentiment_perm3
where NYSE_sentiment = "very bullish"

In [0]:
%sql 
 
-- average Technical Indicator nominal value (20-day Rate of change, 200-day Exponential moving average) when NYSE is Very Bullish 
  
select avg(ROC_20) as Avg_20ROC_Nominal_Value, avg(EMA_200) as Avg_200EMA_Nominal_Value 
from NYSE_sentiment_perm3
where NYSE_sentiment = "very bullish"

In [0]:
%sql 
 
-- average Technical Indicator nominal value (20-day Rate of change, 200-day Exponential moving average) when NYSE is Very Bearish 
  
select avg(ROC_20) as Avg_20ROC_Nominal_Value, avg(EMA_200) as Avg_200EMA_Nominal_Value 
from NYSE_sentiment_perm3
where NYSE_sentiment = "very bearish"

##Queries: Commodity Markets insights

In [0]:
%sql
 
-- NYSE sentiment count when Oil value percentage growth was greater than 5% 
 
select count(oil) as Oil_count, NYSE_sentiment 
from NYSE_sentiment_perm3 
where oil > 0.05 
group by NYSE_sentiment 
order by Oil_count desc

###Queries: Foreign Currency insights

In [0]:
%sql
 
-- average currency Percentage growth (USA, Canada, British, Japan, China) when NYSE is Very Bullish
 
select avg(Dollar_index)*(100) as Avg_US_dlr_grwth, avg(Canadian_dlr)*(100) as Avg_Can_dlr_grwth, avg(British_Pound)*(100) as Avg_Brtsh_Pnd_grwth, avg(Japan_Yen)*(100) as Avg_Jap_Yen_grwth, avg(Chinese_Yuan)*(100) as Avg_Chns_Yuan_grwth 
from NYSE_sentiment_perm3 
where NYSE_sentiment = "very bullish"

In [0]:
%sql
 
-- average currency Percentage growth (USA, Canada, British, Japan, China) when NYSE is Very Bearish
 
select avg(Dollar_index)*(100) as Avg_US_dlr_grwth, avg(Canadian_dlr)*(100) as Avg_Can_dlr_grwth, avg(British_Pound)*(100) as Avg_Brtsh_Pnd_grwth, avg(Japan_Yen)*(100) as Avg_Jap_Yen_grwth, avg(Chinese_Yuan)*(100) as Avg_Chns_Yuan_grwth 
from NYSE_sentiment_perm3 
where NYSE_sentiment = "very bearish" 

###CorrelationTest

In [0]:
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
from pyspark.ml.feature import VectorAssembler
from pyspark.ml.stat import Correlation
 
#Considering only numerical columns
col_labels = ['Volume',
 'ROC_20',
 'EMA_200',
 'Oil',
 'Gold',
 'Copper-F',
 'Wheat-F',
 'London_exg',
 'German_exg-F',
 'Japan_exg-F',
 'Dollar_index',
 'Canadian_dlr',
 'British_Pound',
 'Japan_Yen',
 'Chinese_Yuan',
 'NYSE'] 
 
myGraph = df
vector_col = "corr_features"
assembler = VectorAssembler(inputCols=columns, outputCol=vector_col)
 
myGraph_vector = assembler.transform(myGraph).select(vector_col)
 
matrix = Correlation.corr(myGraph_vector, vector_col).collect()[0][0]
corrmatrix = matrix.toArray().tolist()
print(corrmatrix)
 

In [0]:
def heatmap(data, row_labels, col_labels, ax=None,
            cbar_kw=None, cbarlabel="", **kwargs):
    """
    Create a heatmap from a numpy array and two lists of labels.
 
    Parameters
    ----------
    data
        A 2D numpy array of shape (M, N).
    row_labels
        A list or array of length M with the labels for the rows.
    col_labels
        A list or array of length N with the labels for the columns.
    ax
        A `matplotlib.axes.Axes` instance to which the heatmap is plotted.  If
        not provided, use current axes or create a new one.  Optional.
    cbar_kw
        A dictionary with arguments to `matplotlib.Figure.colorbar`.  Optional.
    cbarlabel
        The label for the colorbar.  Optional.
    **kwargs
        All other arguments are forwarded to `imshow`.
    """
 
    if ax is None:
        ax = plt.gca()
 
    if cbar_kw is None:
        cbar_kw = {}
 
    # Plot the heatmap
    im = ax.imshow(data, **kwargs)
 
    # Create colorbar
    cbar = ax.figure.colorbar(im, ax=ax, **cbar_kw)
    cbar.ax.set_ylabel(cbarlabel, rotation=-90, va="bottom")
 
    # Show all ticks and label them with the respective list entries.
    ax.set_xticklabels(['']+col_labels)
    ax.set_yticklabels(['']+row_labels)
    #ax.set_xticks(np.arange(data), labels=col_labels)
    #ax.set_yticks(np.arange(data), labels=row_labels)
 
    # Let the horizontal axes labeling appear on top.
    ax.tick_params(top=True, bottom=False,
                   labeltop=True, labelbottom=False)
 
    # Rotate the tick labels and set their alignment.
    plt.setp(ax.get_xticklabels(), rotation=-30, ha="right",
             rotation_mode="anchor")
    
 
    # Turn spines off and create white grid.
    ax.spines[:].set_visible(False)
 
    #ax.set_xticks(np.arange(data+1)-.5, minor=True)
    #ax.set_yticks(np.arange(data+1)-.5, minor=True)
    ax.grid(which="minor", color="w", linestyle='-', linewidth=3)
    ax.tick_params(which="minor", bottom=False, left=False)
 
    return im, cbar

In [0]:
def annotate_heatmap(im, data=None, valfmt="{x:.2f}",
                     textcolors=("black", "white"),
                     threshold=None, **textkw):
    """
    A function to annotate a heatmap.
 
    Parameters
    ----------
    im
        The AxesImage to be labeled.
    data
        Data used to annotate.  If None, the image's data is used.  Optional.
    valfmt
        The format of the annotations inside the heatmap.  This should either
        use the string format method, e.g. "$ {x:.2f}", or be a
        `matplotlib.ticker.Formatter`.  Optional.
    textcolors
        A pair of colors.  The first is used for values below a threshold,
        the second for those above.  Optional.
    threshold
        Value in data units according to which the colors from textcolors are
        applied.  If None (the default) uses the middle of the colormap as
        separation.  Optional.
    **kwargs
        All other arguments are forwarded to each call to `text` used to create
        the text labels.
    """
 
    if not isinstance(data, (list, np.ndarray)):
        data = im.get_array()
 
    # Normalize the threshold to the images color range.
    if threshold is not None:
        threshold = im.norm(threshold)
    else:
        threshold = im.norm(data.max())/2.
 
    # Set default alignment to center, but allow it to be
    # overwritten by textkw.
    kw = dict(horizontalalignment="center",
              verticalalignment="center")
    kw.update(textkw)
 
    # Get the formatter in case a string is supplied
    if isinstance(valfmt, str):
        valfmt = mpl.ticker.StrMethodFormatter(valfmt)
 
    # Loop over the data and create a `Text` for each "pixel".
    # Change the text's color depending on the data.
    texts = []
    for i in range(data.shape[0]):
        for j in range(data.shape[1]):
            kw.update(color=textcolors[int(im.norm(data[i, j]) > threshold)])
            text = im.axes.text(j, i, valfmt(data[i, j], None), **kw)
            texts.append(text)
 
    return texts

In [0]:
fig, ax = plt.subplots(figsize=(12, 8))
 
im, cbar = heatmap(corrmatrix, col_labels, col_labels, ax=ax,
                   cmap="coolwarm", cbarlabel="Range")
texts = annotate_heatmap(im, valfmt=" {x:.2f} ")
 
fig.tight_layout()
plt.title("NYSE Correlation -with all columns")
plt.show()

###Building Decision Tree Model

In [0]:
# Selecting the dependent and the independent variables that are identified as most useful attributes to make predictions
 
data=df.select(['Volume', 'ROC_20', 'EMA_200', 'Oil', 'Gold', 'Copper-F', 'Wheat-F',
 'London_exg', 'German_exg-F', 'Japan_exg-F', 'Dollar_index', 'Canadian_dlr', 'British_Pound',
 'Japan_Yen', 'Chinese_Yuan', 'NYSE_sentiment'])

In [0]:
data=data.dropna()

In [0]:
# Create a 70-30 train test split
 
train_data,test_data=data.randomSplit([0.7,0.3])

###Building the Decision Tree Classifier

In [0]:
# Import the required libraries
 
from pyspark.ml.classification import DecisionTreeClassifier
from pyspark.ml.feature import VectorAssembler,StringIndexer
from pyspark.ml import Pipeline

In [0]:
# Use StringIndexer to convert the categorical columns to hold numerical data
NYSE_sentiment_indexer = StringIndexer(inputCol='NYSE_sentiment',outputCol='NYSE_sentiment_index',handleInvalid='keep')

In [0]:
# Vector assembler is used to create a vector of input features
 
assembler = VectorAssembler(inputCols=['Volume', 'ROC_20', 'EMA_200', 'Oil', 'Gold', 'Copper-F', 'Wheat-F',
 'London_exg', 'German_exg-F', 'Japan_exg-F', 'Dollar_index', 'Canadian_dlr', 'British_Pound',
 'Japan_Yen', 'Chinese_Yuan'],
                            outputCol="features")

In [0]:
dt_model = DecisionTreeClassifier(labelCol='NYSE_sentiment_index',maxBins=50000)

In [0]:
pipe = Pipeline(stages=[NYSE_sentiment_indexer,assembler,dt_model])

In [0]:
fit_model=pipe.fit(train_data)

In [0]:
# Store the results in a dataframe
 
results = fit_model.transform(test_data)

In [0]:
results.select(['NYSE_sentiment_index','prediction']).show()

In [0]:
results.show()

###Evaluating the model

In [0]:
from pyspark.ml.evaluation import MulticlassClassificationEvaluator

In [0]:
ACC_evaluator = MulticlassClassificationEvaluator(
    labelCol="NYSE_sentiment_index", predictionCol="prediction", metricName="accuracy")

In [0]:
accuracy = ACC_evaluator.evaluate(results)

In [0]:
print("The accuracy of the decision tree classifier is {}".format(accuracy))

### Building Logistic Regression model

###Data Preprocessing

In [0]:
# Selecting the dependent and the independent variables that are identified as most useful attributes to make predictions
data = df

In [0]:
data=data.dropna()

In [0]:
# Create a 70-30 train test split
train_data,test_data=data.randomSplit([0.7,0.3])

### Importing relevant pyspark libraries and preparing data for its input

In [0]:
# Import the required libraries

from pyspark.ml.classification import LogisticRegression
from pyspark.ml.classification import GBTClassifier
from pyspark.ml.evaluation import MulticlassClassificationEvaluator
from pyspark.ml.feature import VectorAssembler,StringIndexer ,OneHotEncoder
from pyspark.ml import Pipeline
from sklearn.metrics import confusion_matrix

In [0]:
# Use StringIndexer to convert the categorical columns to hold numerical data

NYSE_sentiment_index = StringIndexer(inputCol='NYSE_sentiment',outputCol='NYSE_sentiment_index',handleInvalid='keep')

In [0]:
# Vector assembler is used to create a vector of input features

assembler = VectorAssembler(inputCols=['Volume', 'ROC_20', 'EMA_200', 'Oil', 'Gold', 'Copper-F', 'Wheat-F', 'London_exg', 'German_exg-F', 'Japan_exg-F', 'Dollar_index', 'Canadian_dlr', 'British_Pound', 'Japan_Yen', 'Chinese_Yuan', 'NYSE'],
                            outputCol="features")

In [0]:
# Create an object for the Logistic Regression model

lr_model = LogisticRegression(labelCol='NYSE_sentiment_index')

In [0]:
# Pipeline is used to pass the data through indexer and assembler simultaneously. Also, it helps to pre-rocess the test data
# in the same way as that of the train data.

lr_pipe = Pipeline(stages=[NYSE_sentiment_index, assembler, lr_model])

In [0]:
lr_fit_model = lr_pipe.fit(train_data)

In [0]:
# Store the results in a dataframe

lr_results = lr_fit_model.transform(test_data)

In [0]:
lr_results.select(['NYSE_sentiment_index','prediction']).show()

###Evaluating the Logistic Regression model

#####  1. Accuracy

In [0]:
from pyspark.ml.evaluation import MulticlassClassificationEvaluator

In [0]:
lr_Accuracy_evaluator = MulticlassClassificationEvaluator(predictionCol='prediction',labelCol='NYSE_sentiment_index',metricName='accuracy')

In [0]:
lr_Accuracy = lr_Accuracy_evaluator.evaluate(lr_results)

In [0]:
print("Accuracy of the LR model is {}".format(lr_Accuracy))

####Confusion matrix

In [0]:
lr_y_true = lr_results.select("NYSE_sentiment_index")
lr_y_true = lr_y_true.toPandas()

lr_y_pred = lr_results.select("prediction")
lr_y_pred = lr_y_pred.toPandas()

lr_cnf_matrix = confusion_matrix(lr_y_true, lr_y_pred)
print("Below is the confusion matrix \n {}".format(lr_cnf_matrix))

###Random Forest

In [0]:
from pyspark.ml.classification import RandomForestClassifier

In [0]:
rf_model = RandomForestClassifier(featuresCol = 'features', labelCol = 'NYSE_sentiment_index')

In [0]:
# Pipeline is used to pass the data through indexer and assembler simultaneously. Also, it helps to pre-rocess the test data
# in the same way as that of the train data.

rf_pipe = Pipeline(stages=[NYSE_sentiment_index, assembler, rf_model])

In [0]:
rf_fit_model = rf_pipe.fit(train_data)

In [0]:
# Store the results in a dataframe

rf_results = rf_fit_model.transform(test_data)

In [0]:
rf_results.select(['NYSE_sentiment_index','prediction']).show()

###Evaluating the RF model

In [0]:
rf_Accuracy_evaluator = MulticlassClassificationEvaluator(predictionCol='prediction',labelCol='NYSE_sentiment_index',metricName='accuracy')

In [0]:
rf_Accuracy = rf_Accuracy_evaluator.evaluate(rf_results)

In [0]:
print("Accuracy of the RF model is {}".format(lr_Accuracy))

###Confusion Matrix

In [0]:
rf_y_true = rf_results.select("NYSE_sentiment_index")
rf_y_true = rf_y_true.toPandas()

rf_y_pred = rf_results.select("prediction")
rf_y_pred = rf_y_pred.toPandas()

rf_cnf_matrix = confusion_matrix(rf_y_true, rf_y_pred)
print("Below is the confusion matrix \n {}".format(rf_cnf_matrix))

###Conclusion: 
Based on the above 3 models, we see that Accuracy from Ramdon Forest and Logistic Regression. 

Also, we see that:
1. London_exg and German_exg impact NYSE the most and have about 67% and 66% direct affect.
Followed by Copper at 42% and oil at 26%.
2. Surprisingly Dollar Index and Canadian Dollar have negative impact of about 25% and Gold has NO impact at all.